// @cs-au-dk/jelly/lib/analysis/listeners.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenListener = void 0;
var TokenListener;
(function (TokenListener) {
    TokenListener[TokenListener["CALL_BASE"] = 0] = "CALL_BASE";
    TokenListener[TokenListener["CALL_CALLEE"] = 1] = "CALL_CALLEE";
    TokenListener[TokenListener["CALL_EXTERNAL"] = 2] = "CALL_EXTERNAL";
    TokenListener[TokenListener["READ_BASE"] = 3] = "READ_BASE";
    TokenListener[TokenListener["READ_BASE_DYNAMIC"] = 4] = "READ_BASE_DYNAMIC";
    TokenListener[TokenListener["READ_GETTER"] = 5] = "READ_GETTER";
    TokenListener[TokenListener["READ_GETTER2"] = 6] = "READ_GETTER2";
    TokenListener[TokenListener["READ_GETTER_THIS"] = 7] = "READ_GETTER_THIS";
    TokenListener[TokenListener["READ_GETTER_THIS2"] = 8] = "READ_GETTER_THIS2";
    TokenListener[TokenListener["WRITE_BASE"] = 9] = "WRITE_BASE";
    TokenListener[TokenListener["WRITE_BASE_DYNAMIC"] = 10] = "WRITE_BASE_DYNAMIC";
    TokenListener[TokenListener["WRITE_SETTER"] = 11] = "WRITE_SETTER";
    TokenListener[TokenListener["WRITE_SETTER_THIS"] = 12] = "WRITE_SETTER_THIS";
    TokenListener[TokenListener["WRITE_OBJECT_PATTERN_REST"] = 13] = "WRITE_OBJECT_PATTERN_REST";
    TokenListener[TokenListener["WRITE_OBJECT_PATTERN_REST_PROPERTIES"] = 14] = "WRITE_OBJECT_PATTERN_REST_PROPERTIES";
    TokenListener[TokenListener["WRITE_ARRAY_PATTERN_REST"] = 15] = "WRITE_ARRAY_PATTERN_REST";
    TokenListener[TokenListener["WRITE_ARRAY_PATTERN_REST_ARRAY"] = 16] = "WRITE_ARRAY_PATTERN_REST_ARRAY";
    TokenListener[TokenListener["WRITE_REQUIRE_EXTENSIONS"] = 17] = "WRITE_REQUIRE_EXTENSIONS";
    TokenListener[TokenListener["IMPORT_BASE"] = 18] = "IMPORT_BASE";
    TokenListener[TokenListener["EXPORT_BASE"] = 19] = "EXPORT_BASE";
    TokenListener[TokenListener["ANCESTORS"] = 20] = "ANCESTORS";
    TokenListener[TokenListener["READ_ANCESTORS"] = 21] = "READ_ANCESTORS";
    TokenListener[TokenListener["WRITE_ANCESTORS"] = 22] = "WRITE_ANCESTORS";
    TokenListener[TokenListener["PACKAGE_NEIGHBORS"] = 23] = "PACKAGE_NEIGHBORS";
    TokenListener[TokenListener["CLASS_FIELD"] = 24] = "CLASS_FIELD";
    TokenListener[TokenListener["EXTENDS"] = 25] = "EXTENDS";
    TokenListener[TokenListener["READ_ITERATOR_VALUE"] = 26] = "READ_ITERATOR_VALUE";
    TokenListener[TokenListener["OBJECT_SPREAD"] = 27] = "OBJECT_SPREAD";
    TokenListener[TokenListener["CALL_PROMISE_EXECUTOR"] = 28] = "CALL_PROMISE_EXECUTOR";
    TokenListener[TokenListener["CALL_PROMISE_RESOLVE"] = 29] = "CALL_PROMISE_RESOLVE";
    TokenListener[TokenListener["CALL_PROMISE_ONFULFILLED"] = 30] = "CALL_PROMISE_ONFULFILLED";
    TokenListener[TokenListener["CALL_PROMISE_ONREJECTED"] = 31] = "CALL_PROMISE_ONREJECTED";
    TokenListener[TokenListener["CALL_PROMISE_ONFINALLY"] = 32] = "CALL_PROMISE_ONFINALLY";
    TokenListener[TokenListener["MAKE_PROMISE_RESOLVE"] = 33] = "MAKE_PROMISE_RESOLVE";
    TokenListener[TokenListener["MAKE_PROMISE_REJECT"] = 34] = "MAKE_PROMISE_REJECT";
    TokenListener[TokenListener["MAKE_PROMISE_ALL"] = 35] = "MAKE_PROMISE_ALL";
    TokenListener[TokenListener["MAKE_PROMISE_ALLSETTLED"] = 36] = "MAKE_PROMISE_ALLSETTLED";
    TokenListener[TokenListener["MAKE_PROMISE_ANY"] = 37] = "MAKE_PROMISE_ANY";
    TokenListener[TokenListener["MAKE_PROMISE_RACE"] = 38] = "MAKE_PROMISE_RACE";
    TokenListener[TokenListener["AWAIT"] = 39] = "AWAIT";
    TokenListener[TokenListener["JSX_ELEMENT"] = 40] = "JSX_ELEMENT";
    TokenListener[TokenListener["NATIVE_INVOKE_CALLBACK"] = 41] = "NATIVE_INVOKE_CALLBACK";
    TokenListener[TokenListener["NATIVE_INVOKE_CALLBACK2"] = 42] = "NATIVE_INVOKE_CALLBACK2";
    TokenListener[TokenListener["NATIVE_INVOKE_CALL_APPLY2"] = 43] = "NATIVE_INVOKE_CALL_APPLY2";
    TokenListener[TokenListener["NATIVE_INVOKE_CALL_APPLY3"] = 44] = "NATIVE_INVOKE_CALL_APPLY3";
    TokenListener[TokenListener["NATIVE_ASSIGN_PROPERTIES"] = 45] = "NATIVE_ASSIGN_PROPERTIES";
    TokenListener[TokenListener["NATIVE_ASSIGN_PROPERTIES2"] = 46] = "NATIVE_ASSIGN_PROPERTIES2";
    TokenListener[TokenListener["NATIVE_ASSIGN_PROPERTIES3"] = 47] = "NATIVE_ASSIGN_PROPERTIES3";
    TokenListener[TokenListener["NATIVE_OBJECT_CREATE"] = 48] = "NATIVE_OBJECT_CREATE";
    TokenListener[TokenListener["NATIVE_OBJECT_DEFINE_PROPERTY"] = 49] = "NATIVE_OBJECT_DEFINE_PROPERTY";
    TokenListener[TokenListener["NATIVE_OBJECT_DEFINE_PROPERTIES"] = 50] = "NATIVE_OBJECT_DEFINE_PROPERTIES";
    TokenListener[TokenListener["NATIVE_ASSIGN_ITERATOR_MAP_VALUE_PAIRS"] = 51] = "NATIVE_ASSIGN_ITERATOR_MAP_VALUE_PAIRS";
    TokenListener[TokenListener["NATIVE_ASSIGN_BASE_ARRAY_ARRAY_VALUE_TO_ARRAY"] = 52] = "NATIVE_ASSIGN_BASE_ARRAY_ARRAY_VALUE_TO_ARRAY";
    TokenListener[TokenListener["NATIVE_RETURN_PROTOTYPE_OF"] = 53] = "NATIVE_RETURN_PROTOTYPE_OF";
    TokenListener[TokenListener["NATIVE_SET_PROTOTYPE_OF"] = 54] = "NATIVE_SET_PROTOTYPE_OF";
})(TokenListener || (exports.TokenListener = TokenListener = {}));
